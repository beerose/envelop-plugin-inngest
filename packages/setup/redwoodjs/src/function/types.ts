import type { SetupInngestFunctionOptions } from './command';

export type ExportedType = {
  name: string;
  type: string;
  operationType: 'query' | 'mutation';
};

export interface SetupFunctionTasksOptions extends SetupInngestFunctionOptions {}
