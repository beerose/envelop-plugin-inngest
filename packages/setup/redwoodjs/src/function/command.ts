import type Yargs from 'yargs';

interface BaseOptions {
  cwd: string | undefined;
}

export interface ForceOptions extends BaseOptions {
  force: boolean;
}

export interface SetupInngestFunctionOptions extends ForceOptions {
  name: string;
  type: 'background' | 'scheduled' | 'delayed' | 'step';
}

export const description = 'Set up an Inngest function';

export const builder = (yargs: Yargs.Argv) => {
  return yargs
    .epilogue(
      `Note: To launch the Inngest dev server, from the terminal run: 
    
  npx inngest-cli@latest dev -u http://localhost:8911/inngest    
  `
    )
    .positional('name', { type: 'string', description: 'Name of the function to setup' })
    .option('type', {
      alias: 't',
      type: 'string',
      choices: ['background', 'scheduled', 'delayed', 'step'],
      description: 'Type of Inngest function to setup',
    })
    .option('force', {
      alias: 'f',
      default: false,
      description: 'Overwrite existing files',
      type: 'boolean',
    });
};

export const handler = async (options: any) => {
  const { handler } = await import('./handler');
  return handler(options);
};
