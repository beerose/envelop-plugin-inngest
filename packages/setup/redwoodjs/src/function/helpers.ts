import path from 'path';
import camelcase from 'camelcase';
import fs from 'fs-extra';
import humanize from 'humanize-string';
import template from 'lodash.template';
import { paramCase } from 'param-case';
import { getPaths, writeFile } from '@redwoodjs/cli-helpers';
import type { ExistingFiles } from '@redwoodjs/cli-helpers';
import type { ExportedType, SetupFunctionTasksOptions } from './types';

export const getWebGraphQLTypeDefsFile = () => {
  const GRAPHQL_TYPES_PATH = path.join(getPaths().web.types, 'graphql.d.ts');
  return fs.readFileSync(GRAPHQL_TYPES_PATH, 'utf-8');
};

export const getExportedQueryAndMutationTypes = (graphQLTypeDefs?: string): ExportedType[] => {
  const fileContents = graphQLTypeDefs || getWebGraphQLTypeDefsFile();

  const exportedTypes: ExportedType[] = [];

  // Extract exported types
  const typeRegex =
    /export\s+type\s+(\w+)\s+=\s+{[^}]*__typename\?:\s+['"](?:Query|Mutation)['"][^}]*}/g;

  let match;
  while ((match = typeRegex.exec(fileContents)) !== null) {
    if (!['Mutation', 'Query'].includes(match[1])) {
      const name = match[1];
      const type = match[0];
      const operationType = type.includes("__typename?: 'Query'") ? 'query' : 'mutation';
      exportedTypes.push({ name, type, operationType });
    }
  }

  // Filter exported types with __typename of Query or Mutation
  const filteredTypes = exportedTypes.filter(type => type.operationType === 'query' || 'mutation');

  return filteredTypes.sort((a, b) => {
    if (a.name < b.name) return -1;
    if (a.name > b.name) return 1;
    return 0;
  });
};

export const getNamesForFile = (options: SetupFunctionTasksOptions) => {
  const name = options.name;
  const functionName = camelcase(name);
  const humanizedName = humanize(name);
  const eventFromName = paramCase(name);

  let eventPrefix = 'event';
  let eventName = `${eventPrefix}/${eventFromName}`;

  if (options.graphql && options.type !== 'scheduled' && options.eventName) {
    eventPrefix = 'graphql';
    eventName = `${eventPrefix}/${paramCase(options.eventName)}.${options.operationType}`;
  }

  return { functionName, humanizedName, eventName };
};

export const renderFunctionTemplate = (options: SetupFunctionTasksOptions) => {
  const { eventName, functionName, humanizedName } = getNamesForFile(options);

  const compiled = template(
    fs
      .readFileSync(
        path.resolve(__dirname, '..', '..', 'templates', 'function', `${options.type}.ts.template`),
        'utf-8',
      )
      .toString(),
  );

  const rendered = compiled({ name: options.name, eventName, functionName, humanizedName });

  return { filename: functionName, rendered };
};

export const writeFunctionFile = (
  filename: string,
  rendered: string,
  existingFiles: ExistingFiles,
) => {
  const SRC_INNGEST_PATH = path.join(getPaths().api.src, 'jobs', 'inngest');

  writeFile(path.join(SRC_INNGEST_PATH, `${filename}.ts`), rendered, {
    existingFiles,
  });
};
