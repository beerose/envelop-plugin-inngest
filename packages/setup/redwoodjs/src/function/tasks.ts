import path from 'path';
import camelcase from 'camelcase';
import execa from 'execa';
import fs from 'fs-extra';
import humanize from 'humanize-string';
import * as jscodeshift from 'jscodeshift/src/Runner';
import { Listr } from 'listr2';
import template from 'lodash.template';
import { paramCase } from 'param-case';
import { getPaths, writeFile } from '@redwoodjs/cli-helpers';
import type { ExistingFiles } from '@redwoodjs/cli-helpers';
import type { SetupInngestFunctionOptions } from './command';

export interface SetupFunctionTasksOptions extends SetupInngestFunctionOptions {}

const getNamesForFile = (name: string) => {
  const functionName = camelcase(name);
  const humanizedName = humanize(name);
  const eventName = paramCase(name).replace(/-/g, '.');

  return { functionName, humanizedName, eventName };
};

const SRC_INNGEST_PATH = path.join(getPaths().api.src, 'inngest');

const renderFunctionTemplate = (name: string, type: string) => {
  const { eventName, functionName, humanizedName } = getNamesForFile(name);

  const compiled = template(
    fs
      .readFileSync(
        path.resolve(__dirname, '..', '..', 'templates', 'function', `${type}.ts.template`),
        'utf-8',
      )
      .toString(),
  );

  const rendered = compiled({ name, eventName, functionName, humanizedName });

  return { filename: functionName, rendered };
};

const writeFunctionFile = (filename: string, rendered: string, existingFiles: ExistingFiles) => {
  writeFile(path.join(SRC_INNGEST_PATH, `${filename}.ts`), rendered, {
    existingFiles,
  });
};

export const tasks = (options: SetupFunctionTasksOptions) => {
  const existingFiles = options.force ? 'OVERWRITE' : 'FAIL';

  return new Listr(
    [
      {
        title: `Create a ${options.type} Inngest function named ${options.name} ...`,
        task: () => {
          const { filename, rendered } = renderFunctionTemplate(options.name, options.type);
          writeFunctionFile(filename, rendered, existingFiles);
        },
      },
      {
        title: `Modify the Inngest handler to register the new ${options.type} Inngest function function  named ${options.name}...`,
        task: async () => {
          const { functionName } = getNamesForFile(options.name);

          const SRC_INNGEST_HANDLER_FILE = path.join(getPaths().api.functions, 'inngest.ts');
          const SRC_INNGEST_CODEMOD_FILE = path.join(
            __dirname,
            '..',
            'add-function-to-inngest-handler-codemod.js',
          );

          const defaultJscodeshiftOpts = {
            verbose: 0,
            dry: false,
            print: false,
            babel: false,
            extensions: 'js',
            ignorePattern: '**/node_modules/**',
            ignoreConfig: [],
            runInBand: false,
            silent: false,
            parser: 'ts',
            parserConfig: {},
            failOnError: true,
            stdin: false,
          };

          try {
            await jscodeshift.run(SRC_INNGEST_CODEMOD_FILE, [SRC_INNGEST_HANDLER_FILE], {
              ...defaultJscodeshiftOpts,
              functionName,
            });
          } catch (e: any) {
            // eslint-disable-next-line no-console
            console.error('Failed to modify the GraphQL handler', e.message);
          }
        },
      },
      {
        title: `Lint and Prettify added files ...`,
        task: async () => {
          const SRC_INNGEST_HANDLER_FILE = path.join(getPaths().api.functions, 'inngest.ts');

          execa.commandSync(
            `yarn rw lint --fix ${SRC_INNGEST_HANDLER_FILE}`,
            // eslint-disable-next-line dot-notation
            process.env['RWJS_CWD']
              ? {
                  // eslint-disable-next-line dot-notation
                  cwd: process.env['RWJS_CWD'],
                }
              : {},
          );
        },
      },
    ],
    { rendererOptions: {} },
  );
};
