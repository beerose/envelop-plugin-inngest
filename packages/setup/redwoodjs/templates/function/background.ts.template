import { inngest } from 'src/lib/inngest'

/**
 * To test this ${functionName} function, send an event to the Inngest API with the following payload:
 *
 * {
 *   "name": "${eventName}",
 *   "data": {}
 *   "user": {}
 * }
 *
 */
const ${functionName} = inngest.createFunction(
  { name: '${name}' },
  { event: '${eventName}' },

  // This function will be called every time an event payload is received
  // When implementing your function, you have access to the { event, step }.
  async ({ event }) => {
    return {
      event,
      body: 'Background function ${humanizedName} ran.',
      runAt: Date.now().toString(),
    }
  }
)

export default ${functionName}
