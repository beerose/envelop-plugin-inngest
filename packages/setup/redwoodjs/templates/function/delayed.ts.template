import { inngest } from 'src/lib/inngest'

/**
 * To test this ${functionName} function, send an event to the Inngest API with the following payload:
 *
 * {
 *   "name": "event/${eventName}",
 *   "data": {}
 *   "user": {}
 * }
 *
 */
const ${functionName} = inngest.createFunction(
  { name: '${name}' },
  { event: 'event/${eventName}' },
  async ({ event, step }) => {
    await step.sleep('5s')

    return {
      event,
      body: 'Delayed function ${humanizedName} waited for 5 seconds',
    }
  }
)

export default ${functionName}
