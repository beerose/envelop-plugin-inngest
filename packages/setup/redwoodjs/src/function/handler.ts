import fs from 'fs';
import path from 'path';
import prompts from 'prompts';
import { colors, getPaths } from '@redwoodjs/cli-helpers';
import { tasks as setupFunctionTasks } from './tasks';
import type { SetupFunctionTasksOptions } from './tasks';

interface ErrorWithExitCode extends Error {
  exitCode?: number;
}

type ExportedType = {
  name: string;
  type: string;
  operationType: 'query' | 'mutation';
};

function isErrorWithExitCode(e: unknown): e is ErrorWithExitCode {
  return typeof (e as ErrorWithExitCode)?.exitCode !== 'undefined';
}

function getExportedQueryAndMutationTypes(filePath: string): ExportedType[] {
  const fileContents = fs.readFileSync(filePath, 'utf-8');

  const exportedTypes: ExportedType[] = [];

  // Extract exported types
  const typeRegex =
    /export\s+type\s+(\w+)\s+=\s+{[^}]*__typename\?:\s+['"](?:Query|Mutation)['"][^}]*}/g;

  let match;
  while ((match = typeRegex.exec(fileContents)) !== null) {
    if (!['Mutation', 'Query'].includes(match[1])) {
      const name = match[1];
      const type = match[0];
      const operationType = type.includes("__typename?: 'Query'") ? 'query' : 'mutation';
      exportedTypes.push({ name, type, operationType });
    }
  }

  // Filter exported types with __typename of Query or Mutation
  const filteredTypes = exportedTypes.filter(type => type.operationType === 'query' || 'mutation');

  return filteredTypes.sort((a, b) => {
    if (a.name < b.name) return -1;
    if (a.name > b.name) return 1;
    return 0;
  });
}

export const handler = async ({
  cwd,
  force,
  name,
  type,
  graphql,
  eventName,
  operationType,
}: SetupFunctionTasksOptions) => {
  let functionType = type;
  eventName = name;

  // Prompt to select what type if not specified
  if (!functionType) {
    const response = await prompts({
      type: 'select',
      name: 'functionType',
      choices: [
        {
          value: 'background',
          title: 'Background',
          description: 'Create a background function triggered by an event',
        },
        {
          value: 'scheduled',
          title: 'Scheduled',
          description: 'Create a scheduled function to run at a specific time',
        },
        {
          value: 'delayed',
          title: 'Delayed',
          description: 'Create a delayed function to run after a specified duration',
        },
        {
          value: 'step',
          title: 'Multi-Step',
          description:
            'Create a multi-step function to safely coordinate between events, delay execution for hours or days, and consider previous steps and incoming events',
        },
      ],
      message: 'What type of Inngest function would you like to create?',
    });

    functionType = response.functionType;
  }

  // If GraphQL is specified, the scheduled function does not have an event name.
  if (graphql && functionType !== 'scheduled') {
    const GRAPHQL_TYPES_PATH = path.join(getPaths().web.types, 'graphql.d.ts');

    const operationTypesForEvent = getExportedQueryAndMutationTypes(GRAPHQL_TYPES_PATH);

    const graphqlOperationChoices = operationTypesForEvent.map(op => ({
      value: op,
      title: op.name,
      description: `Create a function for the ${op.operationType} ${op.name}`,
    }));

    const response = await prompts({
      type: 'select',
      name: 'graphQLOperation',
      choices: graphqlOperationChoices,
      message: 'What GraphQL operation event should your function handle?',
    });

    // eslint-disable-next-line no-console
    console.debug(response.graphQLOperation.name, 'Make function for this event');

    operationType = response.graphQLOperation.operationType;
    eventName = response.graphQLOperation.name;
  }

  const tasks = setupFunctionTasks({
    cwd,
    force,
    name,
    type: functionType,
    graphql,
    eventName,
    operationType,
  });

  try {
    await tasks.run();
  } catch (e) {
    if (e instanceof Error) {
      // eslint-disable-next-line no-console
      console.error(colors.error(e.message));
    } else {
      // eslint-disable-next-line no-console
      console.error(colors.error('Unknown error when running yargs tasks'));
    }

    if (isErrorWithExitCode(e)) {
      process.exit(e.exitCode);
    }

    process.exit(1);
  }
};
