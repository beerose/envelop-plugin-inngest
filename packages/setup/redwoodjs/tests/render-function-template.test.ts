import type { SetupFunctionTasksOptions } from '../src/function//types';
import { renderFunctionTemplate } from '../src/function/helpers';

describe('Render function templates', () => {
  describe('standard inngest events', () => {
    it('renders a background function', () => {
      const options: SetupFunctionTasksOptions = {
        name: 'OrderCreated',
        type: 'background',
        eventName: 'OrderCreated',
        graphql: false,
        force: false,
        cwd: __dirname,
      };

      const { rendered } = renderFunctionTemplate(options);

      expect(rendered).toMatchSnapshot();
    });

    it('renders a delayed function', () => {
      const options: SetupFunctionTasksOptions = {
        name: 'AfterCheckout',
        type: 'delayed',
        eventName: 'AfterCheckout',
        graphql: false,
        force: false,
        cwd: __dirname,
      };

      const { rendered } = renderFunctionTemplate(options);

      expect(rendered).toMatchSnapshot();
    });

    it('renders a scheduled function', () => {
      const options: SetupFunctionTasksOptions = {
        name: 'test',
        type: 'scheduled',
        eventName: 'cronJob',
        graphql: false,
        force: false,
        cwd: __dirname,
      };

      const { rendered } = renderFunctionTemplate(options);

      expect(rendered).toMatchSnapshot();
    });

    it('renders a step function', () => {
      const options: SetupFunctionTasksOptions = {
        name: 'OnboardUser',
        type: 'step',
        eventName: 'UserSignUp',
        graphql: false,
        force: false,
        cwd: __dirname,
      };

      const { rendered } = renderFunctionTemplate(options);

      expect(rendered).toMatchSnapshot();
    });

    it('renders a fan-out function', () => {
      const options: SetupFunctionTasksOptions = {
        name: 'SendNewsletter',
        type: 'fan-out',
        eventName: 'RegisterUser',
        graphql: false,
        force: false,
        cwd: __dirname,
      };

      const { rendered } = renderFunctionTemplate(options);

      expect(rendered).toMatchSnapshot();
    });
  });

  describe('graphql triggered events', () => {
    describe('by queries', () => {
      it('renders a background function', () => {
        const options: SetupFunctionTasksOptions = {
          name: 'GetPosts',
          type: 'background',
          eventName: 'FindPosts',
          graphql: true,
          operationType: 'query',
          force: false,
          cwd: __dirname,
        };

        const { rendered } = renderFunctionTemplate(options);

        expect(rendered).toMatchSnapshot();
      });

      it('renders a background function', () => {
        const options: SetupFunctionTasksOptions = {
          name: 'GetPosts',
          type: 'background',
          eventName: 'FindPosts',
          graphql: true,
          operationType: 'query',
          force: false,
          cwd: __dirname,
        };

        const { rendered } = renderFunctionTemplate(options);

        expect(rendered).toMatchSnapshot();
      });

      it('renders a delayed function', () => {
        const options: SetupFunctionTasksOptions = {
          name: 'GetPostBySlug',
          type: 'delayed',
          eventName: 'GetPostBySlug',
          graphql: true,
          operationType: 'query',
          force: false,
          cwd: __dirname,
        };

        const { rendered } = renderFunctionTemplate(options);

        expect(rendered).toMatchSnapshot();
      });

      it('renders a scheduled function', () => {
        const options: SetupFunctionTasksOptions = {
          name: 'GetAllPosts',
          type: 'scheduled',
          eventName: 'FindAllPosts',
          graphql: true,
          operationType: 'query',
          force: false,
          cwd: __dirname,
        };

        const { rendered } = renderFunctionTemplate(options);

        expect(rendered).toMatchSnapshot();
      });

      it('renders a step function', () => {
        const options: SetupFunctionTasksOptions = {
          name: 'GetPost',
          type: 'step',
          eventName: 'FindPostById',
          graphql: true,
          operationType: 'query',
          force: false,
          cwd: __dirname,
        };

        const { rendered } = renderFunctionTemplate(options);

        expect(rendered).toMatchSnapshot();
      });
    });

    describe('by mutations', () => {
      it('renders a background function', () => {
        const options: SetupFunctionTasksOptions = {
          name: 'OnPostUpdate',
          type: 'background',
          eventName: 'UpdatePostById',
          graphql: true,
          operationType: 'mutation',
          force: false,
          cwd: __dirname,
        };

        const { rendered } = renderFunctionTemplate(options);

        expect(rendered).toMatchSnapshot();
      });

      it('renders a delayed function', () => {
        const options: SetupFunctionTasksOptions = {
          name: 'OnPostDelete',
          type: 'delayed',
          eventName: 'DeletePostById',
          graphql: true,
          operationType: 'mutation',
          force: false,
          cwd: __dirname,
        };

        const { rendered } = renderFunctionTemplate(options);

        expect(rendered).toMatchSnapshot();
      });

      it('renders a step function', () => {
        const options: SetupFunctionTasksOptions = {
          name: 'OnPostEdit',
          type: 'step',
          eventName: 'EditPostById',
          graphql: true,
          operationType: 'mutation',
          force: false,
          cwd: __dirname,
        };

        const { rendered } = renderFunctionTemplate(options);

        expect(rendered).toMatchSnapshot();
      });

      it('renders a fan-out function', () => {
        const options: SetupFunctionTasksOptions = {
          name: 'WhenNewPost',
          type: 'fan-out',
          eventName: 'AddPost',
          graphql: true,
          operationType: 'mutation',
          force: false,
          cwd: __dirname,
        };

        const { rendered } = renderFunctionTemplate(options);

        expect(rendered).toMatchSnapshot();
      });
    });
  });
});
