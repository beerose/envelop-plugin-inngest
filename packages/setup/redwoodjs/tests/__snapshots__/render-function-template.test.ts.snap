// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render function templates graphql triggered events by mutations renders a background function 1`] = `
"import { inngest } from 'src/lib/inngest'

/**
 * To test this onPostUpdate function, send an event to the Inngest API with the following payload:
 *
 * {
 *   "name": "graphql/update-post-by-id.mutation",
 *   "data": {}
 *   "user": {}
 * }
 *
 */
const onPostUpdate = inngest.createFunction(
  { name: 'OnPostUpdate' },
  { event: 'graphql/update-post-by-id.mutation' },

  // This function will be called every time an event payload is received
  // When implementing your function, you have access to the { event, step }.
  async ({ event }) => {
    return {
      event,
      body: 'Background function On post update ran.',
      runAt: Date.now().toString(),
    }
  }
)

export default onPostUpdate
"
`;

exports[`Render function templates graphql triggered events by mutations renders a delayed function 1`] = `
"import { inngest } from 'src/lib/inngest'

/**
 * To test this onPostDelete function, send an event to the Inngest API with the following payload:
 *
 * {
 *   "name": "graphql/delete-post-by-id.mutation",
 *   "data": {}
 *   "user": {}
 * }
 *
 */
const onPostDelete = inngest.createFunction(
  { name: 'On post delete' },
  { event: 'graphql/delete-post-by-id.mutation' },
  async ({ event, step }) => {
    await step.sleep('5s')

    return {
      event,
      body: 'Delayed function On post delete waited for 5 seconds',
    }
  }
)

export default onPostDelete
"
`;

exports[`Render function templates graphql triggered events by mutations renders a fan-out function 1`] = `
"import { inngest } from 'src/lib/inngest'

/**
 * To test this whenNewPost function make sure your app and dev server is running
 */
const whenNewPost = inngest.createFunction(
  { name: 'When new post' }, // The name of your function, used for observability.
  { cron: 'TZ=America/New_York * * * * *' }, // The cron syntax for the function. TZ= is optional.

  // This function will be called on the schedule defined in the cron property.
  // When implementing your function, you have access to the { event, step }.
  async ({ step }) => {
    // Fetch all users
    const users = await step.run('Fetch users', async () => {
      return fetchUsers()
    })

    // For each user, send us an event.  Inngest supports batches of events
    // as long as the entire payload is less than 512KB.
    const events = users.map((user) => {
      return {
        name: 'app/weekly-email-activity.send',
        data: {
          ...user,
        },
        user,
      }
    })

    // Send all events to Inngest, which triggers any functions listening to
    // the given event names.
    await step.sendEvent(events)

    // Return the number of users triggered.
    return { count: users.length }
  }
)

export const whenNewPostFanOut = inngest.createFunction(
  { name: 'Weekly activity: send email' },
  { event: 'app/weekly-email-activity.send' },
  async ({ event, step }) => {
    const data = await step.run('Load user data', async () => {
      return loadUserData(event.data.id)
    })

    await step.run('Email user', async () => {
      return await sendEmail(event.data.user, data)
    })
  }
)

const fetchUsers = () => {
  return [
    { user: { id: 1, name: 'Alice', email: 'alice@example.com' } },
    { user: { id: 2, name: 'Bob', email: 'bob@example.com' } },
    { user: { id: 3, name: 'Carol', email: 'carol@example.com' } },
    { user: { id: 4, name: 'David', email: 'david@example.com' } },
    { user: { id: 5, name: 'Eve', email: 'eve@example.com' } },
  ]
}

const loadUserData = async (id) => {
  return fetchUsers().find((user) => user.user.id === id)
}

const sendEmail = async (user, data) => {
  const message = \`Sending email to \${user.name} at \${user.email} ...\`
  return { message, user, data }
}
"
`;

exports[`Render function templates graphql triggered events by mutations renders a step function 1`] = `
"import { inngest } from 'src/lib/inngest'

/**
 * To test this onPostEdit function, send an event to the Inngest API with the following payload:
 *
 * {
 *   "name": "graphql/edit-post-by-id.mutation",
 *   "data": { "user": { id: 123 } }
 *   "user": {}
 * }
 *
 *
 * If you do not send another event named "app/order.created", then next step will not run.
 *
 *
 * If you do send a "app/order.created" event, make sure the data.user.id matches the id 
 * used in the \`app/graphql/edit-post-by-id.mutation\` event sent to complete the event handling.
 *
 * {
 *   "name": "app/order.created",
 *   "data": { "user": { "id": 123 } }
 *   "user": {}
 * }
 * 
 */
const onPostEdit = inngest.createFunction(
  { name: 'On post edit' },
  { event: 'graphql/edit-post-by-id.mutation' },
  async ({ event, step }) => {
    // Send the user a welcome email
    await step.run('Send welcome email', () => {
      return {
        event,
        body: 'Hello World!',
        userId: event.data?.user?.id,
      }
    })

    // Wait for the user to create an order, by waiting and
    // matching on another event
    const order = await step.waitForEvent('app/order.created', {
      match: 'data.user.id',
      timeout: '20s',
    })

    if (!order) {
      // User didn't create an order within 20 seconds;
      // send them an activation email
      await step.run('Send activation email', async () => {
        return {
          event,
          body: 'Mail sent!',
          userId: event.data?.user?.id,
        }
      })
    }
  }
)

export default onPostEdit
"
`;

exports[`Render function templates graphql triggered events by queries renders a background function 1`] = `
"import { inngest } from 'src/lib/inngest'

/**
 * To test this getPosts function, send an event to the Inngest API with the following payload:
 *
 * {
 *   "name": "graphql/find-posts.query",
 *   "data": {}
 *   "user": {}
 * }
 *
 */
const getPosts = inngest.createFunction(
  { name: 'GetPosts' },
  { event: 'graphql/find-posts.query' },

  // This function will be called every time an event payload is received
  // When implementing your function, you have access to the { event, step }.
  async ({ event }) => {
    return {
      event,
      body: 'Background function Get posts ran.',
      runAt: Date.now().toString(),
    }
  }
)

export default getPosts
"
`;

exports[`Render function templates graphql triggered events by queries renders a background function 2`] = `
"import { inngest } from 'src/lib/inngest'

/**
 * To test this getPosts function, send an event to the Inngest API with the following payload:
 *
 * {
 *   "name": "graphql/find-posts.query",
 *   "data": {}
 *   "user": {}
 * }
 *
 */
const getPosts = inngest.createFunction(
  { name: 'GetPosts' },
  { event: 'graphql/find-posts.query' },

  // This function will be called every time an event payload is received
  // When implementing your function, you have access to the { event, step }.
  async ({ event }) => {
    return {
      event,
      body: 'Background function Get posts ran.',
      runAt: Date.now().toString(),
    }
  }
)

export default getPosts
"
`;

exports[`Render function templates graphql triggered events by queries renders a delayed function 1`] = `
"import { inngest } from 'src/lib/inngest'

/**
 * To test this getPostBySlug function, send an event to the Inngest API with the following payload:
 *
 * {
 *   "name": "graphql/get-post-by-slug.query",
 *   "data": {}
 *   "user": {}
 * }
 *
 */
const getPostBySlug = inngest.createFunction(
  { name: 'Get post by slug' },
  { event: 'graphql/get-post-by-slug.query' },
  async ({ event, step }) => {
    await step.sleep('5s')

    return {
      event,
      body: 'Delayed function Get post by slug waited for 5 seconds',
    }
  }
)

export default getPostBySlug
"
`;

exports[`Render function templates graphql triggered events by queries renders a scheduled function 1`] = `
"import { inngest } from 'src/lib/inngest'

/**
 * To test this getAllPosts function make sure your app and dev server is running
 */
const getAllPosts = inngest.createFunction(
  { name: 'Get all posts' }, // The name of your function, used for observability.
  { cron: 'TZ=America/New_York * * * * *' }, // The cron syntax for the function. TZ= is optional.

  // This function will be called on the schedule defined in the cron property.
  // When implementing your function, you have access to the { event, step }.
  async () => {
    return {
      message: \`Scheduled Get all posts function ran.\`,
      runAt: Date.now().toString(),
    }
  }
)

export default getAllPosts
"
`;

exports[`Render function templates graphql triggered events by queries renders a step function 1`] = `
"import { inngest } from 'src/lib/inngest'

/**
 * To test this getPost function, send an event to the Inngest API with the following payload:
 *
 * {
 *   "name": "graphql/find-post-by-id.query",
 *   "data": { "user": { id: 123 } }
 *   "user": {}
 * }
 *
 *
 * If you do not send another event named "app/order.created", then next step will not run.
 *
 *
 * If you do send a "app/order.created" event, make sure the data.user.id matches the id 
 * used in the \`app/graphql/find-post-by-id.query\` event sent to complete the event handling.
 *
 * {
 *   "name": "app/order.created",
 *   "data": { "user": { "id": 123 } }
 *   "user": {}
 * }
 * 
 */
const getPost = inngest.createFunction(
  { name: 'Get post' },
  { event: 'graphql/find-post-by-id.query' },
  async ({ event, step }) => {
    // Send the user a welcome email
    await step.run('Send welcome email', () => {
      return {
        event,
        body: 'Hello World!',
        userId: event.data?.user?.id,
      }
    })

    // Wait for the user to create an order, by waiting and
    // matching on another event
    const order = await step.waitForEvent('app/order.created', {
      match: 'data.user.id',
      timeout: '20s',
    })

    if (!order) {
      // User didn't create an order within 20 seconds;
      // send them an activation email
      await step.run('Send activation email', async () => {
        return {
          event,
          body: 'Mail sent!',
          userId: event.data?.user?.id,
        }
      })
    }
  }
)

export default getPost
"
`;

exports[`Render function templates standard inngest events renders a background function 1`] = `
"import { inngest } from 'src/lib/inngest'

/**
 * To test this orderCreated function, send an event to the Inngest API with the following payload:
 *
 * {
 *   "name": "event/order-created",
 *   "data": {}
 *   "user": {}
 * }
 *
 */
const orderCreated = inngest.createFunction(
  { name: 'OrderCreated' },
  { event: 'event/order-created' },

  // This function will be called every time an event payload is received
  // When implementing your function, you have access to the { event, step }.
  async ({ event }) => {
    return {
      event,
      body: 'Background function Order created ran.',
      runAt: Date.now().toString(),
    }
  }
)

export default orderCreated
"
`;

exports[`Render function templates standard inngest events renders a delayed function 1`] = `
"import { inngest } from 'src/lib/inngest'

/**
 * To test this afterCheckout function, send an event to the Inngest API with the following payload:
 *
 * {
 *   "name": "event/after-checkout",
 *   "data": {}
 *   "user": {}
 * }
 *
 */
const afterCheckout = inngest.createFunction(
  { name: 'After checkout' },
  { event: 'event/after-checkout' },
  async ({ event, step }) => {
    await step.sleep('5s')

    return {
      event,
      body: 'Delayed function After checkout waited for 5 seconds',
    }
  }
)

export default afterCheckout
"
`;

exports[`Render function templates standard inngest events renders a fan-out function 1`] = `
"import { inngest } from 'src/lib/inngest'

/**
 * To test this sendNewsletter function make sure your app and dev server is running
 */
const sendNewsletter = inngest.createFunction(
  { name: 'Send newsletter' }, // The name of your function, used for observability.
  { cron: 'TZ=America/New_York * * * * *' }, // The cron syntax for the function. TZ= is optional.

  // This function will be called on the schedule defined in the cron property.
  // When implementing your function, you have access to the { event, step }.
  async ({ step }) => {
    // Fetch all users
    const users = await step.run('Fetch users', async () => {
      return fetchUsers()
    })

    // For each user, send us an event.  Inngest supports batches of events
    // as long as the entire payload is less than 512KB.
    const events = users.map((user) => {
      return {
        name: 'app/weekly-email-activity.send',
        data: {
          ...user,
        },
        user,
      }
    })

    // Send all events to Inngest, which triggers any functions listening to
    // the given event names.
    await step.sendEvent(events)

    // Return the number of users triggered.
    return { count: users.length }
  }
)

export const sendNewsletterFanOut = inngest.createFunction(
  { name: 'Weekly activity: send email' },
  { event: 'app/weekly-email-activity.send' },
  async ({ event, step }) => {
    const data = await step.run('Load user data', async () => {
      return loadUserData(event.data.id)
    })

    await step.run('Email user', async () => {
      return await sendEmail(event.data.user, data)
    })
  }
)

const fetchUsers = () => {
  return [
    { user: { id: 1, name: 'Alice', email: 'alice@example.com' } },
    { user: { id: 2, name: 'Bob', email: 'bob@example.com' } },
    { user: { id: 3, name: 'Carol', email: 'carol@example.com' } },
    { user: { id: 4, name: 'David', email: 'david@example.com' } },
    { user: { id: 5, name: 'Eve', email: 'eve@example.com' } },
  ]
}

const loadUserData = async (id) => {
  return fetchUsers().find((user) => user.user.id === id)
}

const sendEmail = async (user, data) => {
  const message = \`Sending email to \${user.name} at \${user.email} ...\`
  return { message, user, data }
}
"
`;

exports[`Render function templates standard inngest events renders a scheduled function 1`] = `
"import { inngest } from 'src/lib/inngest'

/**
 * To test this test function make sure your app and dev server is running
 */
const test = inngest.createFunction(
  { name: 'Test' }, // The name of your function, used for observability.
  { cron: 'TZ=America/New_York * * * * *' }, // The cron syntax for the function. TZ= is optional.

  // This function will be called on the schedule defined in the cron property.
  // When implementing your function, you have access to the { event, step }.
  async () => {
    return {
      message: \`Scheduled Test function ran.\`,
      runAt: Date.now().toString(),
    }
  }
)

export default test
"
`;

exports[`Render function templates standard inngest events renders a step function 1`] = `
"import { inngest } from 'src/lib/inngest'

/**
 * To test this onboardUser function, send an event to the Inngest API with the following payload:
 *
 * {
 *   "name": "event/onboard-user",
 *   "data": { "user": { id: 123 } }
 *   "user": {}
 * }
 *
 *
 * If you do not send another event named "app/order.created", then next step will not run.
 *
 *
 * If you do send a "app/order.created" event, make sure the data.user.id matches the id 
 * used in the \`app/event/onboard-user\` event sent to complete the event handling.
 *
 * {
 *   "name": "app/order.created",
 *   "data": { "user": { "id": 123 } }
 *   "user": {}
 * }
 * 
 */
const onboardUser = inngest.createFunction(
  { name: 'Onboard user' },
  { event: 'event/onboard-user' },
  async ({ event, step }) => {
    // Send the user a welcome email
    await step.run('Send welcome email', () => {
      return {
        event,
        body: 'Hello World!',
        userId: event.data?.user?.id,
      }
    })

    // Wait for the user to create an order, by waiting and
    // matching on another event
    const order = await step.waitForEvent('app/order.created', {
      match: 'data.user.id',
      timeout: '20s',
    })

    if (!order) {
      // User didn't create an order within 20 seconds;
      // send them an activation email
      await step.run('Send activation email', async () => {
        return {
          event,
          body: 'Mail sent!',
          userId: event.data?.user?.id,
        }
      })
    }
  }
)

export default onboardUser
"
`;
